### PR TITLE
Add instance volumes

### LINKED ISSUE #1894: [Feature]: Instance (aka host, local) volumes (aka mounts)
### Problem

Currently, the only way to share any data between runs is to use volumes, which are supported only by a small number of backends and not supported on SSH fleets at all.

If it was possible to specify an arbitrary bind mount in the run configuration — `/path/on/fleet/instance:/path/inside/run/container` (akin Docker `-v` [bind mounts](https://docs.docker.com/engine/storage/bind-mounts/)), it would be possible to at least partially work around this limitation.

Possible use cases:

* Cache shared between runs (any runs, not necessarily the same run configuration), e.g., `~/.cache/huggingface`, `~/.cache/pip`. As it’s acceptable for the software cache to be empty, it’s okay if the host directory doesn’t exist or contains irrelevant data.
* Persistent(-ish) storage. If the user is sure that nobody reuses their fleet instance, then they can use the instance mount point as some sort of persistent storage. The lifetime of such storage is bound to the lifetime of the instance.
* Network mounts that are not yet supported by `dstack` (NFS, SMB, etc.) on SSH fleet instances.

### Solution

_No response_

### Workaround

_No response_

### Would you like to help us implement this feature by sending a PR?

Yes

### PR DESCRIPTION
**Pull request by un-def**
Closes: https://github.com/dstackai/dstack/issues/1894

### CODE REVIEW

**Review by jvstme — APPROVED** [review]
That was fast!

File: src/dstack/_internal/core/models/configurations.py [hunk 1]
```diff
@@ -143,6 +150,14 @@ def convert_python(cls, v, values) -> Optional[PythonVersion]:
             return PythonVersion(v)
         return v
 
+    @validator("volumes", each_item=True)
+    def convert_volumes(cls, v) -> Union[VolumeMountPoint, InstanceMountPoint]:
+        if isinstance(v, str):
+            if v.startswith("/"):
+                return InstanceMountPoint.parse(v)
+            return VolumeMountPoint.parse(v)
```
Comment (r4victor) [inline src/dstack/_internal/core/models/configurations.py]: So now we support `volume-name:mount_path`? Please mention such things in the PR description or in the docs.

Comment (r4victor) [reply src/dstack/_internal/core/models/configurations.py]: Is it a good idea at all? User may try to use relative paths for instance mounts but will get an error about volumes?

**Review by r4victor — APPROVED** [review]

**Comment by r4victor** [comment]
Updating docs as a part of the PR makes it much easier to review.

File: src/dstack/_internal/core/models/configurations.py (hunk 1 above)
Comment (un-def) [reply src/dstack/_internal/core/models/configurations.py]: I've improved mount point type detection, if the left part contains a slash in any position, we are sure it's an instance mount point: https://github.com/dstackai/dstack/pull/1895/commits/8e0396e37ac5291dcc8a71469bb49053773c49e9#diff-6b2463188801b3ca761b5edf2026af9f3c226b150fb67b5c7fad1665bdeacc9eR96-R99

The only case when the ambiguity is still present is when the user tries to mount a relative instance path `foo` with the `foo:/path/to` syntax — we assume it's a volume mount and fail with `Volume foo not found`.

Not sure if the verbose message [here](https://github.com/dstackai/dstack/blob/d8b11a3381a3d0a9073eb95843e402acc1a405ca/src/dstack/_internal/server/services/runs.py#L635) (something like “to use a network volume {name_or_rel_path}, create it with the volume configuration, to mount a local path, specify an absolute path `/{name_or_rel_path}`”) would greatly improve user experience, for example, it the user uses the full form syntax:

```yaml
volumes:
  - name: foo
    path: /volume
```

it's clear that they tries to mount a network volume, but the error message would still suggest a local volume fix as well. What do you think?

Comment (r4victor) [reply src/dstack/_internal/core/models/configurations.py]: > we assume it's a volume mount and fail with 

I think it's ok. No need for a different error message.

**Review by jvstme — APPROVED** [review]

File: docs/docs/concepts/volumes.md [hunk 2]
```diff
@@ -112,19 +121,99 @@ $ dstack delete -f vol.dstack.yaml
...
+```yaml
+type: dev-environment
+# A name of the dev environment
+name: vscode-vol
+
+ide: vscode
+
+# Map the instance path to any container path
+volumes:
+  - instance_path: /mnt/volume
```
Comment (peterschmidt85) [inline docs/docs/concepts/volumes.md:157]: Unsure about `instance_path` and `path`.
Perhaps `instance_path` and `container_path` may be more consistent.
Also, I personally suggest promoting everywhere the short syntax:

```
volumes:
  - <instance path>:<container path>
 ```

File: src/dstack/_internal/core/models/volumes.py [hunk 3]
```diff
@@ -115,8 +115,8 @@ def parse(cls, v: str) -> Self:
 
 
 class InstanceMountPoint(CoreModel):
-    instance_path: Annotated[str, Field(description="The path on the instance (host)")]
-    path: Annotated[str, Field(description="The path in the container")]
+    instance_path: Annotated[str, Field(description="The absolute path on the instance (host)")]
+    path: Annotated[str, Field(description="The absolute path in the container")]
```
Comment (peterschmidt85) [inline src/dstack/_internal/core/models/volumes.py:119]: Does it mean relative paths are forbidden for container paths too?

**Review by peterschmidt85 — APPROVED** [review]

**Pull request merged**